### PR TITLE
Temperature & Redshift

### PR DESCRIPTION
**Pull request by fjebaker**
Added in some old code from before the refactor of GeodesicTracer.jl so that redshift values for the CarterBoyerLindquist method may be calculated. At the moment, this still requires a little bit of boiler plate as I don't want to commit to an API until I've finished looking at some analytic derivations for the general 2nd order case redshift, and had a chance to think about how the dispatch is going to work.

Here's an example script in the mean time:
```julia
using GeodesicRendering
using AccretionGeometry
using CarterBoyerLindquist
using AccretionFormulae

using Plots
gr()

m = CarterMethodBL(M=1.0, a=1.0)

# observer position
u = [0.0, 1000.0, deg2rad(85.0), 0.0]

# disc has inner radius 10, outer radius 50, perpendicular to the spin axis
d = GeometricThinDisc(10.0, 50.0, deg2rad(90.0))

# wrapper around the redshift function
function redshift(m, sol, max_time; kwargs...)
    u = sol.u[end] 
    AccretionFormulae.regular_pdotu_inv(u, sol.prob.p, m)
end

# create and compose the ValueFunction
redshift_vf = (
    # calculate redshift
    ValueFunction(redshift)
    # filter only pixels with r > 9, i.e. on the disc
    ∘ FilterValueFunction((m, sol, max_time; kwargs...) -> sol.u[end][2] > 9, NaN)
    # filter only pixels that terminated early (i.e., intersected with something)
    ∘ ConstValueFunctions.filter_early_term
)

# do the render
img = @time rendergeodesics(
    m, u, 2000.0, 
    d, 
    fov_factor=3.0, abstol=1e-9, reltol=1e-9,
    image_width = 350,
    image_height = 250,
    vf =redshift_vf
)

# plot
heatmap(reverse(img, dims=1))
```

For `CarterMethodBL`, the tolerances need to be a little better, i.e. 1e-9. If scaling the image up, e.g. twice the size (as below), `image_width`, `image_height` and `fov_factor` should be multiplied by the same scaling integer `n`.

Output:
redshift_example

### CODE REVIEW

**Comment by fjebaker** [comment]
## Developing after the refactor

It's probably easiest to use the Julia Pkg `develop` or `dev` command, along with all the benefits they bring, and to create a development environment "seperate" from the actual package you are working on. I would recommend creating a new directory to keep this development environment in -- in this example, `DevEnv/`. 

AccretionFormulae depends on two packages, which we can install in the new environment. To do this, we just need to add the Astro Group Bristol registry:
```julia
(@1.7) julia>] activate .
(DevEnv) pkg> registry add https://github.com/astro-group-bristol/AstroRegistry
```
We can now install the dependencies we need just as we would any other package
```julia
(DevEnv) pkg> add GeodesicRendering CarterBoyerLindquist
```

Next, we want to check out the AccretionFormulae repository for development; we select specifically the `fergus/wip-redshift` branch so we have access to the code in this draft PR:
```
(DevEnv) pkg> dev --local https://github.com/astro-group-bristol/AccretionFormulae.jl#fergus/wip-redshift
```
The `dev` command creates a (local) clone of the desired package, and tells Julia to use this in the environment. Once we have finished developing a `dev` package, we can `free` it to use the registry versions again.

You should now have a directory structure similar to:
```
DevEnv/
├── Manifest.toml
├── Project.toml
├── dev
│   └── AccretionFormulae.jl/
└── script1.jl
```
To make sure everything is working correctly, I would recommend copying the script above into a new file `script1.jl` and make sure you can recreate the above plot. If anything isn't working, tag me in this PR with your issues and I will do my best to help.

Packages you install into the `DevEnv`, such as [Plots.jl](https://docs.juliaplots.org/stable/) or [BenchmarkTools.jl](https://github.com/JuliaCI/BenchmarkTools.jl), are then added to the Package.toml of the development environment, and *not* the repository you are developing, allowing you to keep development tools and project tools seperate :) 

## Development practice

I would strongly recommend using Tim Holy's [Revise.jl](https://timholy.github.io/Revise.jl/stable/) when developing. Install it into `DevEnv` with
```julia
(DevEnv) pkg> add Revise
```
then simply call
```
julia> using Revise
```
in the REPL _before_ `using` any other libraries. This will ensure that if you modify files in e.g. `dev/AccretionFormulae` that they will be reloaded by Revise, and ensures you don't have to keep restarting your REPL whenver you make changes.

This is quite different from how you might have developed in Python before, but lends itself to exploratory work much more succinctly -- hopefully you'll agree!

**Comment by fjebaker** [comment]
Oh, one last thing! DifferentialEquations recently released a substantial update, so it may take several minutes to precompile GeodesicRendering again. The first time you run the script, you may wish to take a short coffee break ;)

**Comment by kyle8mcknight** [comment]
Hi @fjebaker, we're trying to get an image of the observed temperature using your redshift function. We need to take the product of the temperature at each radius (from our function), with the corresponding g factor (as per the Fanton paper). In order to do this we need to find the g factor at each radius, however we're unsure how to obtain g values as a function of radius. It seems from the 'redshift()' function above, the line: 'AccretionFormulae.regular_pdotu_inv(u, sol.prob.p, m)' gives the g factor?  Would you be able to suggest a way to combine these g values with the corresponding values of T, i.e. at the same radius? Finally, once we've got a function producing redshift adjusted temperature values, is a case of inputting this into the 'ValueFunction()', as in the above script? Thanks, Kyle and Dan

**Comment by fjebaker** [comment]
Yes! So this is pushing up against an API choice that I haven't quite settled on yet. I'd like for the `ValueFunction`s to be composable, as they already are in the case of the `FilterValueFunctions`, but I haven't quite worked out how to do this best.

## tl;dr

The easiest thing to do at the moment is to _not_ try and compose your temperature function with the redshift function, but just evaluate it in your own `ValueFunction`. To be more detailed:

All of the information of the geodesic solution is contained in `sol`;
- the position of the intersection is stored in `sol.u[end]`, i.e. the last four-vector in the geodesic path.
- `sol.u[end]` is an array with 4 elements, with index `1` being time, and `2-4` being `r`, `theta`, and `phi` respectively.

You can therefore write your temperature function along these lines:
```julia
function temperature_function(m, sol, max_time; kwargs...)
    # calculate redshift at intersection position
    g = redshift(m, sol, max_time; kwargs...)
    
    # get intersection position
    u = sol.u[end] 
    
    # calculate temperature using `g` and `u`
    temperature = ...
    
    return temperature
end
```
This obviously requires the `redshift` function above. Then, the full `ValueFunction` you'd want to compose is:
```julia
temperature_vf = (
    ValueFunction(temperature_function)
    ∘ FilterValueFunction((m, sol, max_time; kwargs...) -> sol.u[end][2] > 9, NaN)
    ∘ ConstValueFunctions.filter_early_term
)
```
which will filter the pixel values so you're only calculating the temperature for geodesics that actually _intersected_ the disc, and not those which e.g. travelled to infinity or fell into the event horizon.


## Speculative API

At the moment, whatever the previous `ValueFunction` returns is passed to the next (if there is one); so doing something like
```julia
vf = ValueFunction(temperature_function)∘ ValueFunction(redshift)
```
would pass the redshift calculated as the sole argument to (`temperature_function`). This is obviously not desireable, since all information about the geodesic intersection is lost. It may be recovered by ammending
```julia
function redshift(m, sol, max_time; kwargs...)
    u = sol.u[end] 
    g = AccretionFormulae.regular_pdotu_inv(u, sol.prob.p, m)
    (sol, g)
end
```
which now passes `sol` and the redshift value `g` to the temperature function as arguments. Though this works for intended purposes, it doesn't _feel_ generic enough for future applications, where my hope is that complex functions could be chained together from many smaller functions without them having to be "aware" of eachother's output formats.

At the moment, the `kwargs...` argument has been left in for perhaps such a purpose, passing the output of previous `ValueFunction`s to the next via this parameter, but I think the signature is a little strange like this...

If you have a suggestion for how this might elegantly work, in such a way as to allow `ValueFunction`s to string together or exist solely, please let me know!

- The previous implementation used a `px` value, which represented the pixel value computed so far (initialised to `NaN`) which the `ValueFunction`s continously modified, but I didn't like that particularly. 
- Another idea I am considering is having a distinction between `ValueFunction`s which use/modify a previous result (as temperature would), and those which just generate a value, as with `redshift`.

For now just compute the redshift function within your temperature function, and we'll see how that can be decoupled in the future.

**Comment by fjebaker** [comment]
@kyle8mcknight Another point to mention, which I forgot in the above: all of the metric-relevant quantities, i.e. black hole spin and mass, are in the `m` parameter passed into the `ValueFunction`s.

- `m.M` is the black hole mass in gravitational units;
- `m.a` is the black hole spin, again in gravitational units.

I don't believe it is relevant for your needs, but for completeness, the `max_time` parameter is the value of the affine parameter at which the geodesic terminated.

**Comment by dangoh123** [comment]
Thanks for the help @fjebaker. We've tried to make a temperature function based on your advice, but have run into an issue with the following [error](https://pastebin.com/H3TXgQAg). We think this is the same error we were getting when we were initially playing with the mass and spin. Here is our code:

Emission Profile:

```julia
#imports
using Plots

#functions
"""
From Reynolds (2020) (Eq2-4)
Calculates the radius of the inner most stable circular orbit of the a black hole,
with spin a_star, and mass M.
"""
function r_isco(a_star, M)
    
    a = a_star
    r_g = 2*G*M/(c^2)
    z1 = 1+∛(1-a^2)*(∛(1+a)+∛(1-a))
    z2 = √(3*a^2+z1^2)
    if a >= 0
        r_isco = (3+z2-(3-z1)*√(3+z1+2*z2))*r_g
    elseif a < 0
        r_isco = (3+z2+(3-z1)*√(3+z1+2*z2))*r_g
    end
end

"""
From Page & Thorne (1974) (Eq15n)
Calculates the function f at radius r, for a black hole of spin a_star, and mass M, 
which is used to calculate the flux given by its accretion disk.
"""
function f(r, a_star, M)

    R_isco = r_isco(a_star, M)
  
    x = √(r/M)
    x0 = √(R_isco/M)
    x1 = 2*cos((1/3)*(acos(a_star))-(π/3))
    x2 = 2*cos((1/3)*(acos(a_star))+(π/3))
    x3 = -2*cos((1/3)*(acos(a_star)))

    f = (3/(2*M))*(1/(x^2*(x^3-(3*x)+(2*a_star))))*(x-x0-((3/2)*a_star*log(x/x0)) 
            - (3*(x1-a_star)^2)/(x1*(x1-x2)*(x1-x3))*log((x-x1)/(x0-x1)) 
            - (3*(x2-a_star)^2)/(x2*(x2-x1)*(x2-x3))*log((x-x2)/(x0-x2)) 
            - (3*(x3-a_star)^2)/(x3*(x3-x1)*(x3-x2))*log((x-x3)/(x0-x3)))
end

function diss(mdot, r, a_star, M)
    diss = mdot*f(r, a_star, M)/(4*π*r)
end

function mdot(M)
    L_edd = 3e4*L_☼*(M/M_☼)
    Mdot = -L_edd/(c^2*η)
    mdot = 0.1*Mdot
end

function temp(r, a_star, M)
    m_dot = mdot(M)
    temp = (diss(m_dot, r, a_star, M)/σ_SB)^(1/4)
end

"""
From Fanton et al. (1997) (Eq78)
"""
function temp_obs(r, a_star, M, g)
    temp = temp(r, a_star, M)
    temp_obs =  g*temp
end

# constants
G = 6.67e-11
c = 3e8
L_☼ = 3.8e26
M_☼ = 1.99e30
σ_SB = 5.67e-8

η = 0.1
# M = 10*M_☼
# a_star = 0.1

module EmissionProfile

export temp_obs, r_isco

end
```

Script to test:

```julia
using GeodesicRendering
using AccretionGeometry
using CarterBoyerLindquist
using AccretionFormulae
include("emission_profile.jl")
#using emission_profile

using Plots
gr()

m = CarterMethodBL(M=1.0, a=1.0)

# observer position
u = [0.0, 1000.0, deg2rad(85.0), 0.0]
R_isco = r_isco(m.a, m.M)

# disc has inner radius 10, outer radius 50, perpendicular to the spin axis
d = GeometricThinDisc(R_isco, 50.0, deg2rad(90.0))

# wrapper around the redshift function
function redshift(m, sol, max_time; kwargs...)
    u = sol.u[end] 
    AccretionFormulae.regular_pdotu_inv(u, sol.prob.p, m)
end

function temperature(m, sol, max_time; kwargs...)
    g =  redshift(m, sol, max_time: kwargs...)

    u = sol.u[end]

    M = m.M*1.99e30
    a = m.a
    a_star = a/M
    temperature = temp_obs(u[2], a_star, M, g)
end

# create and compose the ValueFunction
redshift_vf = (
    # calculate redshift
    ValueFunction(temperature)
    # filter only pixels with r > 9, i.e. on the disc
    ∘ FilterValueFunction((m, sol, max_time; kwargs...) -> sol.u[end][2] > r_isco, NaN)
    # filter only pixels that terminated early (i.e., intersected with something)
    ∘ ConstValueFunctions.filter_early_term
)

# do the render
img = @time rendergeodesics(
    m, u, 2000.0, 
    d, 
    fov_factor=3.0, abstol=1e-9, reltol=1e-9,
    image_width = 350,
    image_height = 250,
    vf =redshift_vf
)

# plot
heatmap(reverse(img, dims=1))
```

If there's any more information you need to diagnose this, just let us know.

We were also having problems correctly importing the functions from the files, and settled on `include("emission_profile.jl")`, which worked some of the time but doesn't seem right if you have any insight.

**Comment by fjebaker** [comment]
Hi @dangoh123, the error your getting seems to be because you're trying to compare the value of a float to a function name, in
```julia
redshift_vf = (
    ValueFunction(temperature)
    ∘ FilterValueFunction((m, sol, max_time; kwargs...) -> sol.u[end][2] > r_isco, NaN)
    #                                                                      ^^^^^^
    
    ∘ ConstValueFunctions.filter_early_term
)
```
Should this be `R_isco`?

I will try running this code later today if you still have problems!

**Comment by fjebaker** [comment]
When you open a PR with your new code I'll review how to structure your work into the module, and we can resolve the import issues you are having then 👍

**Comment by fjebaker** [comment]
Actually, thinking about that a little more, when you open the PR, can you select this branch as the one to merge into?

**Comment by dangoh123** [comment]
@fjebaker cheers for that, I've opened a PR with all the new things. 

I changed r_isco to R_isco, but that didn't seem to make any difference. These two similar variables exist because Julia was having problems with the variables and functions sharing names if you know anything about that.

**Comment by fjebaker** [comment]
Temperature plots in their current standing for maximally spinning Black Hole:

temp1

temp2

**Comment by fjebaker** [comment]
Resolved differences in the redshift calculations in Carter and 2nd Order methods:

side-by-side-redshift

redshift-difference

Merging to keep future updates clean.

**Comment by dangoh123** [comment]
Hi @fjebaker, just a quick one about the scale of the graph. As you can probably see from any temperature plots, the tick labels on the colourbar go off the plot. eg:
NEW

This is quite annoying, as we can't really figure out if our scale is correct. Preferably, we would be able to have scientific notation on the scale, but so far I haven't been able to find any simple way of doing this. I found a couple of threads ([1](https://discourse.julialang.org/t/scientific-notation-in-plots-jl-colorbar/57793), [2](https://github.com/JuliaPlots/Plots.jl/pull/3346)) discussing this, but they all seemed to require installation of other packages like PyPlot or Plotly, which adds more annoying dependencies, and hasn't been working very well. I haven't managed to find an elegant solution to this, and was wondering if you were very familiar with the syntax of Julia's default Plots package?

**Comment by fjebaker** [comment]
@dangoh123 What I usually do with Plots.jl is divide the data manually, and then put the scaling in the title or somewhere similar:
```julia
img ./= 1e4 # just a guess

heatmap(img)
title!("Temperature x1e4")
```

I've not used Plots.jl to create publication quality plots before, I just use it as a quick tool to view data, create interactions (plotly) and create animations. If you really want fine control over your plots, I would highly recommend [Makie.jl](https://makie.juliaplots.org/stable/) (also by the JuliaPlots community, but uses Cairo as a backend and has a much nicer layout and attribute management system).

**Comment by fjebaker** [comment]
Another option would be to play with the size of the figure (I don't know if this works, but it's an idea):
```julia
hm = heatmap(img)
plot(hm, size=(800, 300))
```

**Comment by dangoh123** [comment]
@fjebaker that first method seemed to do the trick for now, thanks. Hopefully there'll be a more elegant solution for our final plots, but this should get us on the right path.

**Comment by dangoh123** [comment]
> @dangoh123 What I usually do with Plots.jl is divide the data manually, and then put the scaling in the title or somewhere similar:
> 
> ```julia
> img ./= 1e4 # just a guess
> 
> heatmap(img)
> title!("Temperature x1e4")
> ```
> 
> I've not used Plots.jl to create publication quality plots before, I just use it as a quick tool to view data, create interactions (plotly) and create animations. If you really want fine control over your plots, I would highly recommend [Makie.jl](https://makie.juliaplots.org/stable/) (also by the JuliaPlots community, but uses Cairo as a backend and has a much nicer layout and attribute management system).

Just one quick question about the first line. What exactly does the `.` represent in `./=`? I get that the `/=` is a divide equals, but I can't find anything about the `.`.
